Laravel-Backpack/CRUD
# Add support for Eloquent Model events to fields

## WHY

### BEFORE - What was wrong? What was happening before this PR?

For `upload`, `upload_multiple` and `image` fields, you had to resort to Mutators in order to actually upload stuff to the disk, then save the filename to the database.

### AFTER - What is happening after this PR?

You can define stuff to happen when standard Eloquent Model events are triggered. For example, you can do:

```php
// using the ARRAY SYNTAX, define an array of events and closures
CRUD::addField([
    'name' => 'name',
    'events' => [
        'saving' => function ($entry) {
            $entry->name = strtoupper($entry->name);
        },
    ],
]);

// using the FLUENT SYNTAX, you can define the same array
CRUD::field('name')->events([
    'saving' => function ($entry) {
        $entry->name = strtoupper($entry->name);
    },
]);
// BUT you can also use the convenience method "on" to define just ONE event
CRUD::field('name')->on('saving', function ($entry) {
    $entry->name = strtoupper($entry->name);
});
```

It supports [all Eloquent events](https://laravel.com/docs/master/eloquent#events), but... the ones that make sense are just: ~~`retrieved`~~, `creating`, `created`, `updating`, `updated`, `saving`, `saved`, ~~`deleting`~~, ~~`deleted`~~, ~~`restoring`~~, ~~`restored`~~, and ~~`replicating`~~, because those are the events that actually take place inside the Create or Update operation.

You CANNOT register events that happen OUTSIDE this particular operation. I mean, you can, but they won't be run. For example it's no use doing stuff on the `deleting` event inside `setupCreateOperation()`... because `deleting` will never happen there, it will only happen inside the `Delete` operation.

----

You can, however, do this kind of general stuff inside the `setup()` method, which is run inside all operations. There, you can use standard Laravel syntax, for example:

```php
Monster::deleting(function ($entry) {
    // TODO: delete that entry's files from the disk
});
```

This is also a preferred way of doing stuf before / after an entry has been create/updated. We're going to instruct people to do this, instead of overriding the `store()` or `update()` methods. Either in `setup()` or `setupUpdateOperation()` or `setupCreateOperation()`.

## HOW

### How did you achieve that, in technical terms?

Before we actually do the creating/updating inside `store()` and `update()`, we go through all fields, see if they have `events` defined and if so, register those events.

For the general way of adding model events, in `setup()`, we did nothing. It was already working, we just didn't think of it.

### Is it a breaking change or non-breaking change?

non-breaking

### How can we test the before & after?

Inside `CategoryCrudController::setupCreateOperation()` for example, you can do this:
```php
CRUD::field('name')->on('saving', function ($entry) {
    $entry->name = strtoupper($entry->name);
});
```

Before, this did nothing, it just added it to the field definition. 
Now, after you create/update the entry, the name should be uppercased.


## WHAT

What needs to be done next?
- [x] decide if we should use `when('saving', $closure)` or `on('saving', $closure)`... or both;
- [x] test & review
